File: test/ModularPipelines.UnitTests/JsonSerializationTests.cs
using System.Text.Json;
using Microsoft.Extensions.DependencyInjection;
using ModularPipelines.Context;
using ModularPipelines.Engine;
using ModularPipelines.Extensions;
using ModularPipelines.Models;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests;

public class JsonSerializationTests : TestBase
{
    public class Module1 : Module<IDictionary<string, object>?>
    {
        public override async Task<IDictionary<string, object>?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();

            return new Dictionary<string, object>
            {
                ["Foo"] = "Bar",
                ["Hello"] = "world!"
            };
        }
    }

    public class Module2 : Module<IDictionary<string, object>?>
    {
        public override async Task<IDictionary<string, object>?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();

            return new Dictionary<string, object>
            {
                ["Foo2"] = "Bar",
                ["Hello2"] = "world!"
            };
        }
    }

    public class Module3 : Module<IDictionary<string, object>?>
    {
        public override async Task<IDictionary<string, object>?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();

            return new Dictionary<string, object>
            {
                ["Foo3"] = "Bar",
                ["Hello3"] = "world!"
            };
        }
    }

    [Test]
    public async Task Test1()
    {
        var host = await TestPipelineHostBuilder.Create()
            .AddModule<Module1>()
            .BuildHostAsync();

        var pipelineSummary = await host.ExecutePipelineAsync();

        var resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>();
        var module1Result = resultRegistry.GetResult<IDictionary<string, object>?>(typeof(Module1))!;

        // Serialize and deserialize the pipeline summary
        var pipelineJson = JsonSerializer.Serialize(pipelineSummary);
        var deserializedSummary = JsonSerializer.Deserialize<PipelineSummary>(pipelineJson);

        using (Assert.Multiple())
        {
            await Assert.That(pipelineJson).IsNotNull().And.IsNotEmpty();
            await Assert.That(deserializedSummary).IsNotNull();
            await Assert.That(deserializedSummary!.Start).IsEqualTo(pipelineSummary.Start);
            await Assert.That(deserializedSummary.End).IsEqualTo(pipelineSummary.End);
            await Assert.That(deserializedSummary.TotalDuration).IsEqualTo(pipelineSummary.TotalDuration);
            // Modules are not serialized (interface types can't be deserialized)
            await Assert.That(deserializedSummary.Modules).HasCount().EqualTo(0);
        }

        // Verify the module result values
        using (Assert.Multiple())
        {
            await Assert.That(module1Result.Value!["Foo"].ToString()).IsEqualTo("Bar");
            await Assert.That(module1Result.Value!["Hello"].ToString()).IsEqualTo("world!");
            await Assert.That(module1Result.ModuleName).IsEqualTo(typeof(Module1).Name);
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Models/JsonSerializationTests.fs
namespace ModularPipelines.UnitTests.FSharp.Models

open System.Collections.Generic
open System.Text.Json
open Microsoft.Extensions.DependencyInjection
open ModularPipelines.Engine
open ModularPipelines.Extensions
open ModularPipelines.Models
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type private JsonModule1() =
    inherit Module<IDictionary<string, obj>>()
    override _.ExecuteAsync(_, _) =
        task {
            do! System.Threading.Tasks.Task.Yield()
            let d = Dictionary<string, obj>()
            d["Foo"] <- "Bar"
            d["Hello"] <- "world!"
            return d :> IDictionary<string, obj>
        }

type JsonSerializationTests() =
    inherit TestBase()

    [<Test>]
    member _.Test1() = async {
        let! host =
            TestPipelineHostBuilder.Create()
                .AddModule<JsonModule1>()
                .BuildHostAsync()
            |> Async.AwaitTask

        let! pipelineSummary = host.ExecutePipelineAsync() |> Async.AwaitTask

        let resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>()
        let module1Result = resultRegistry.GetResult<IDictionary<string, obj>>(typeof<JsonModule1>)

        let pipelineJson = JsonSerializer.Serialize(pipelineSummary)
        let deserializedSummary = JsonSerializer.Deserialize<PipelineSummary>(pipelineJson)

        do! check(Assert.That(pipelineJson).IsNotNull())
        do! check(Assert.That(pipelineJson).IsNotEmpty())
        do! check(Assert.That(deserializedSummary).IsNotNull())
        do! check(Assert.That(deserializedSummary.Start = pipelineSummary.Start).IsTrue())
        do! check(Assert.That(deserializedSummary.End = pipelineSummary.End).IsTrue())
        do! check(Assert.That(deserializedSummary.TotalDuration = pipelineSummary.TotalDuration).IsTrue())
        do! check(Assert.That(deserializedSummary.Modules |> Seq.isEmpty).IsTrue())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(module1Result.ValueOrDefault["Foo"].ToString()), "Bar"))
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(module1Result.ValueOrDefault["Hello"].ToString()), "world!"))
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(module1Result.ModuleName), typeof<JsonModule1>.Name))
    }
